extra_tests/docs_gen.py

```diff
@@ -0,0 +1,101 @@
+import sys
+import warnings
+from pydoc import ModuleScanner
+
+
+def scan_modules():
+    """taken from the source code of help('modules')
+
+    https://github.com/python/cpython/blob/63298930fb531ba2bb4f23bc3b915dbf1e17e9e1/Lib/pydoc.py#L2178"""
+    modules = {}
+
+    def callback(path, modname, desc, modules=modules):
+        if modname and modname[-9:] == ".__init__":
+            modname = modname[:-9] + " (package)"
+        if modname.find(".") < 0:
+            modules[modname] = 1
+
+    def onerror(modname):
+        callback(None, modname, None)
+
+    with warnings.catch_warnings():
+        # ignore warnings from importing deprecated modules
+        warnings.simplefilter("ignore")
+        ModuleScanner().run(callback, onerror=onerror)
+    return list(modules.keys())
+
+
+def import_module(module_name):
+    import io
+    from contextlib import redirect_stdout
+    # Importing modules causes ('Constant String', 2, None, 4) and
+    # "Hello world!" to be printed to stdout.
+    f = io.StringIO()
+    with warnings.catch_warnings(), redirect_stdout(f):
+        # ignore warnings caused by importing deprecated modules
+        warnings.filterwarnings("ignore", category=DeprecationWarning)
+        try:
+            module = __import__(module_name)
+        except Exception as e:
+            return e
+    return module
+
+
+def is_child(module, item):
+    import inspect
+    item_mod = inspect.getmodule(item)
+    return item_mod is module
+
+
+def traverse(module, names, item):
+    import inspect
+    is_builtin = inspect.ismodule(item) or inspect.isbuiltin(item)
```

Comment:
`inspect.isbuiltin` only deals with `methods/functions`. This won't return docstrings for classes, unfortunately. Probably a `has_doc` variable that checks if its also a class would be better.